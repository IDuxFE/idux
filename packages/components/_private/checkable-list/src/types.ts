/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
