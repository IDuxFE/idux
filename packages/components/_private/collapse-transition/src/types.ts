/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
