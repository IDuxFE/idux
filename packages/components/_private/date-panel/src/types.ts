/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
