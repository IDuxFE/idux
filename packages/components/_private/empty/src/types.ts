/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
