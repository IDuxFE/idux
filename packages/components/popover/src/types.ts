import type { DefineComponent, HTMLAttributes } from 'vue'
import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'

import { IxPropTypes } from '@idux/cdk/utils'
import { ɵTooltipProps } from '@idux/components/tooltip'

export const popoverProps = {
  ...ɵTooltipProps,
  content: IxPropTypes.string,
}

export type PopoverProps = IxInnerPropTypes<typeof popoverProps>
export type PopoverPublicProps = IxPublicPropTypes<typeof popoverProps>
export type PopoverComponent = DefineComponent<HTMLAttributes & typeof popoverProps>
export type PopoverInstance = InstanceType<DefineComponent<PopoverProps>>
