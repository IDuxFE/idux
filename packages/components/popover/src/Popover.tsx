/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { PopoverProps } from './types'

import { Ref, Slots, VNode, computed, defineComponent } from 'vue'

import { ɵHeader } from '@idux/components/_private/header'
import { ɵOverlay } from '@idux/components/_private/overlay'
import { useGlobalConfig } from '@idux/components/config'
import { ɵUseTooltipOverlay } from '@idux/components/tooltip'

import { popoverProps } from './types'

export default defineComponent({
  name: 'IxPopover',
  props: popoverProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('popover')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-popover`)
    const { overlayRef, updatePopper, overlayProps, setVisible } = ɵUseTooltipOverlay(
      props,
      config,
      common,
      mergedPrefixCls,
    )

    expose({ updatePopper })

    const closeIcon = computed(() => props.closeIcon ?? config.closeIcon)

    return () => {
      const prefixCls = mergedPrefixCls.value
      return (
        <ɵOverlay
          ref={overlayRef}
          class={prefixCls}
          {...overlayProps.value}
          v-slots={{
            default: slots.default,
            content: () => renderContent(props, slots, closeIcon, setVisible, prefixCls),
          }}
          transitionName={`${common.prefixCls}-fade`}
        />
      )
    }
  },
})

const renderContent = (
  props: PopoverProps,
  slots: Slots,
  closeIcon: Ref<string | VNode>,
  setVisibility: (visible: boolean) => void,
  prefixCls: string,
) => {
  if (!(slots.header || props.header || slots.content || props.content)) {
    return null
  }

  let cls = `${prefixCls}-wrapper`

  const children: VNode[] = []
  if (slots.header || props.header) {
    children.push(
      <ɵHeader
        size="sm"
        closable={props.closable}
        closeIcon={closeIcon.value}
        header={props.header ?? props.title}
        onClose={() => setVisibility(false)}
        v-slots={{ header: slots.header }}
      />,
    )

    cls += ` ${prefixCls}-with-header`
  }

  if (slots.content || props.content) {
    children.push(<div class={`${prefixCls}-content`}>{slots.content?.() ?? props.content}</div>)
  }

  return <div class={cls}>{children}</div>
}
