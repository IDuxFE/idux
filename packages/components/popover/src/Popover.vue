<template>
  <ix-tooltip v-bind="options" @visibleChange="onVisibleChange">
    <template #title>
      <div class="ix-popover-inner">
        <div v-if="hasTitle" class="ix-popover-title">
          <slot name="title">{{ title }}</slot>
        </div>
        <div class="ix-popover-inner-content">
          <slot name="content">{{ content }}</slot>
        </div>
      </div>
    </template>
    <slot />
  </ix-tooltip>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import { hasSlot } from '@idux/cdk/utils'
import { IxTooltip } from '@idux/components/tooltip'
import { useGlobalConfig } from '@idux/components/config'
import { popoverProps } from './types'

export default defineComponent({
  name: 'IxPopover',
  components: { IxTooltip },
  props: popoverProps,
  emits: ['update:visible'],
  setup(props, { emit, slots }) {
    const options = computed(() => {
      const config = useGlobalConfig('popover')
      return {
        visible: props.visible,
        placement: props.placement ?? config.placement,
        trigger: props.trigger ?? config.trigger,
        showDelay: props.showDelay ?? config.showDelay,
        hideDelay: props.hideDelay ?? config.hideDelay,
        destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
        autoAdjust: props.autoAdjust ?? config.autoAdjust,
        clsPrefix: 'ix-popover',
      }
    })

    const hasTitle = computed(() => {
      return !!(hasSlot(slots, 'title') || props.title)
    })

    const onVisibleChange = (visible: boolean) => {
      emit('update:visible', visible)
    }

    return { options, onVisibleChange, hasTitle }
  },
})
</script>
