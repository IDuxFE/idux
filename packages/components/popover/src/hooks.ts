import type { ComputedRef, WritableComputedRef } from 'vue'
import type { PopoverConfig } from '@idux/components/config'
import type { PopoverProps } from './types'

import { computed, getCurrentInstance } from 'vue'
import { useGlobalConfig } from '@idux/components/config'
import { callEmit } from '@idux/cdk/utils'

export function useConfig(): ComputedRef<PopoverConfig> {
  const config = useGlobalConfig('popover')
  const props = getCurrentInstance()!.props as PopoverProps

  return computed(() => {
    return {
      placement: props.placement ?? config.placement,
      trigger: props.trigger ?? config.trigger,
      showDelay: props.showDelay ?? config.showDelay,
      hideDelay: props.hideDelay ?? config.hideDelay,
      destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
      autoAdjust: props.autoAdjust ?? config.autoAdjust,
    }
  })
}

export function useVisibility(): WritableComputedRef<boolean> {
  const props = getCurrentInstance()!.props as PopoverProps

  return computed({
    get() {
      return props.visible
    },
    set(value: boolean) {
      callEmit(props['onUpdate:visible'], value)
    },
  })
}
