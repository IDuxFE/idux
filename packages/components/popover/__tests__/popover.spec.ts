import { mount, MountingOptions, VueWrapper } from '@vue/test-utils'
import { defineComponent } from 'vue'
import { renderWork, wait } from '@tests'
import IxPopover from '../src/Popover.vue'
import { PopoverProps } from '../src/types'

const TestComponent = defineComponent({
  components: { IxPopover },
  props: ['title', 'content', 'placement', 'visible', 'trigger', 'destroyOnHide'],
  template: `
  <ix-popover :title="title" :content="content" :placement="placement" :visible="visible" :trigger="trigger" :destroy-on-hide="destroyOnHide">
    <template v-if='!!$slots.title' #title><slot name='title'/></template>
    <template v-if="!!$slots.content" #content><slot name="content"/></template>
    <span>Popover will show when it's clicked.</span>
  </ix-popover>
  `,
})

describe('Popover.vue', () => {
  let PopoverMount: (options?: MountingOptions<PopoverProps>) => VueWrapper<InstanceType<typeof TestComponent>>

  beforeEach(() => {
    PopoverMount = options => mount(TestComponent, { ...options })
  })

  afterEach(() => {
    document.body.querySelectorAll('.ix-popover').forEach(value => {
      value.remove()
    })
  })

  renderWork(IxPopover, { props: { content: 'Content' } })

  test('visible work', async () => {
    const wrapper = PopoverMount({ props: { title: 'Title', content: 'Content' } })
    expect((document.body.querySelector('.ix-popover') as HTMLDivElement).style.display).toEqual('none')
    await wrapper.get('span').trigger('click')
    await wait(100)
    expect((document.body.querySelector('.ix-popover') as HTMLDivElement).style.display).toEqual('')
  })
})
