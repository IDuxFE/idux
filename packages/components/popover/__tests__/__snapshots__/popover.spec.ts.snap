// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover header props work 1`] = `"<transition-stub><div class=\\"ix-overlay ix-popover\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px;\\"><div class=\\"ix-popover-wrapper\\"><div class=\\"ix-header ix-header-sm\\"><div class=\\"ix-header-content\\"><span class=\\"ix-header-prefix\\"><i class=\\"ix-icon ix-icon-up\\" role=\\"img\\" aria-label=\\"up\\"></i></span><!----><span class=\\"ix-header-title\\">Title</span><span class=\\"ix-header-sub-title\\">subTitle</span><span class=\\"ix-header-suffix\\"><i class=\\"ix-icon ix-icon-down\\" role=\\"img\\" aria-label=\\"down\\"></i></span></div><!----></div></div><div class=\\"ix-overlay-arrow\\" style=\\"position: absolute;\\"></div></div></transition-stub>"`;

exports[`Popover render work 1`] = `
"<div id=\\"trigger\\">trigger</div>
<!--teleport start-->
<!--teleport end-->"
`;
